### DISCUSSION TITLE
Is there any solution for MQTT QoS 2 message pulish connection drop issue?

### DISCUSSION THREAD

**Discussion by tsigcn** [rabbitmq/rabbitmq-server#11842] [Questions] [answered]
I have met a connection droped issue when publish QoS 2 message to RabbitMQ.

The same as the following:
[https://github.com/rabbitmq/rabbitmq-mqtt/issues/155](https://github.com/rabbitmq/rabbitmq-mqtt/issues/155)

Is there any solution to downgrade QoS2 to QoS1 when publish MQTT messages?

Any advice will be appreciated.

**Comment by michaelklishin** [comment]
@tsigcn QoS 2 is not supported by RabbitMQ. Such messages are treated as QoS 1 ones.

We cannot comment on "dropped connections" without an executable way to reproduce. We do not guess in this community.

**Reply by tsigcn** [reply]
@michaelklishin Sorry for the wrong link to rabbit-mqtt issue #155 before.  I have some information for this issue to reproduce:

My test tools:
1. mosquitto-2.0.18a for windows x86-64
2. MQTT.fx 5.7.0

(1) Output of msquitto:
PS C:\Program Files\mosquitto> .\mosquitto_pub.exe -h 20.205.137.88 -p 1883 -u username -P password -t my.test -m "test for qos 2" -q 2 -d
Client null sending CONNECT
Client null received CONNACK (0)
Client null sending PUBLISH (d0, q2, r0, m1, 'my.test', ... (14 bytes))
Client null received PUBACK (Mid: 1, RC:0)
Error: A network protocol error occurred when communicating with the broker.

(2) Log and snapshot of MQTT.fx:
Profile Name ...................................: New Profile 3
Profile Type ...................................: MQTT Broker
Broker Address .................................: 20.205.137.88
Broker Port ....................................: 1883
Client ID ......................................: MQTT_FX_Client
MQTT Version ...................................: 3.1.1
Use SimpleAuth .................................: true
Username .......................................: mqadm
Password .......................................: 
Use SSL/TLS ....................................: false
SSL/TLS version ................................: TLSv1.2
Trust all Hostnames ............................: false
Use default CA certificate .....................: false
Use CA certificate file ........................: false
Use CA certificate Keystore (JKS) ..............: false
Use Certificate Files...........................: false
Use Certificates from trusted Keystore (JKS) ...: false
Use Key Algorithm ..............................: RSA
Auto Reconnect .................................: false
Connection Timeout .............................: 30
Keepalive Interval .............................: 60
Clean session ..................................: true
Use Last Will & Testament ......................: false
Last Will Destination ..........................: 
Last Will QOS ..................................: 0
Last Will Retained .............................: false

2024-07-26 13:59:46,395  INFO --- MqttFXClient                   : attempt to connect with MQTT v3.1.1
2024-07-26 13:59:47,871  INFO --- MqttFXClientModel              : MqttClient with ID Optional[MQTT_FX_Client] assigned.
2024-07-26 13:59:47,873  INFO --- MqttFXClient                   : connected: CONNACK
2024-07-26 13:59:47,906  INFO --- MqttFXClient                   : MQTT version: 3.1.1
2024-07-26 13:59:47,906  INFO --- MqttFXClient                   : session present: false
2024-07-26 14:00:22,620  INFO --- PublishController              : publish
2024-07-26 14:00:22,629  INFO --- MqttFXClientModel              : attempt to add PublishTopic
2024-07-26 14:00:22,635  INFO --- MqttFXClientModel              : addPublishTopic : Test
2024-07-26 14:00:22,643  INFO --- MqttFXClient                   : Broker disconnected: Broker connection is disconnected PUBACK must not be received for a QoS 2 PUBLISH
2024-07-26 14:00:22,694  INFO --- ScriptsController              : Clear console.


<img width="728" alt="snapshot_mqttfx" src="https://github.com/user-attachments/assets/e36fc9de-8dcb-4bac-bb1a-36d5494adbd0">

**Reply by tsigcn** [reply]
And here is output of "rabbitmqctl status"

Status of node rabbit@RabbitMQ ...
Runtime

OS PID: 2703
OS: Linux
Uptime (seconds): 1661651
Is under maintenance?: false
RabbitMQ version: 3.13.4
RabbitMQ release series support status: see https://www.rabbitmq.com/release-information
Node name: rabbit@RabbitMQ
Erlang configuration: Erlang/OTP 26 [erts-14.2.5.1] [source] [64-bit] [smp:2:2] [ds:2:2:10] [async-threads:1] [jit:ns]
Crypto library: OpenSSL 3.0.2 15 Mar 2022
Erlang processes: 438 used, 1048576 limit
Scheduler run queue: 1
Cluster heartbeat timeout (net_ticktime): 60

Plugins

Enabled plugin file: /etc/rabbitmq/enabled_plugins
Enabled plugins:

 * rabbitmq_web_mqtt
 * rabbitmq_mqtt
 * rabbitmq_management
 * rabbitmq_management_agent
 * rabbitmq_web_dispatch
 * amqp_client
 * cowboy
 * cowlib
 * oauth2_client
 * jose

Data directory

Node data directory: /var/lib/rabbitmq/mnesia/rabbit@RabbitMQ
Raft data directory: /var/lib/rabbitmq/mnesia/rabbit@RabbitMQ/quorum/rabbit@RabbitMQ

Config files


Log file(s)

 * /var/log/rabbitmq/rabbit@RabbitMQ.log
 * <stdout>

Alarms

(none)

Memory

Total memory used: 0.1738 gb
Calculation strategy: rss
Memory high watermark setting: 0.4 of available memory, computed to: 3.3046 gb

reserved_unallocated: 0.098 gb (56.39 %)
code: 0.0398 gb (22.88 %)
other_proc: 0.0242 gb (13.92 %)
other_system: 0.0231 gb (13.32 %)
other_ets: 0.0038 gb (2.16 %)
atom: 0.0016 gb (0.9 %)
plugins: 0.0013 gb (0.77 %)
binary: 0.001 gb (0.59 %)
metrics: 0.0006 gb (0.33 %)
mgmt_db: 0.0002 gb (0.13 %)
metadata_store: 0.0002 gb (0.11 %)
msg_index: 0.0001 gb (0.07 %)
mnesia: 0.0001 gb (0.04 %)
metadata_store_ets: 0.0 gb (0.02 %)
quorum_ets: 0.0 gb (0.01 %)
connection_other: 0.0 gb (0.0 %)
quorum_queue_procs: 0.0 gb (0.0 %)
quorum_queue_dlx_procs: 0.0 gb (0.0 %)
stream_queue_procs: 0.0 gb (0.0 %)
stream_queue_replica_reader_procs: 0.0 gb (0.0 %)
connection_readers: 0.0 gb (0.0 %)
connection_writers: 0.0 gb (0.0 %)
connection_channels: 0.0 gb (0.0 %)
queue_procs: 0.0 gb (0.0 %)
queue_slave_procs: 0.0 gb (0.0 %)
stream_queue_coordinator_procs: 0.0 gb (0.0 %)
allocated_unused: 0.0 gb (0.0 %)

File Descriptors

Total: 0, limit: 32671
Sockets: 0, limit: 29401

Free Disk Space

Low free disk space watermark: 0.05 gb
Free disk space: 28.8944 gb

Totals

Connection count: 0
Queue count: 0
Virtual host count: 1

Listeners

Interface: [::], port: 25672, protocol: clustering, purpose: inter-node and CLI tool communication
Interface: [::], port: 5672, protocol: amqp, purpose: AMQP 0-9-1 and AMQP 1.0
Interface: [::], port: 15672, protocol: http, purpose: HTTP API
Interface: [::], port: 1883, protocol: mqtt, purpose: MQTT
Interface: [::], port: 15675, protocol: http/web-mqtt, purpose: MQTT over WebSockets

**Reply by michaelklishin** [reply]
We need RabbitMQ version and server logs.

**Reply by tsigcn** [reply]
And here is output of "rabbitmqctl environment"

Application environment of node rabbit@RabbitMQ ...
[{amqp10_common,[]},
 {amqp_client,
     [{gen_server_call_timeout,130000},
      {prefer_ipv6,false},
      {ssl_options,[]},
      {writer_gc_threshold,1000000000}]},
 {asn1,[]},
 {aten,
     [{detection_threshold,0.99},
      {heartbeat_interval,100},
      {poll_interval,5000},
      {scaling_factor,1.5}]},
 {compiler,[]},
 {cowboy,[]},
 {cowlib,[]},
 {credentials_obfuscation,[{enabled,true}]},
 {crypto,[{fips_mode,false},{rand_cache_size,896}]},
 {cuttlefish,[]},
 {enough,[]},
 {erts,[]},
 {gen_batch_server,[]},
 {horus,[{skip_collection_froms_apps,[]}]},
 {inets,[]},
 {jose,[{crypto_fallback,true},{pbes2_count_maximum,10000}]},
 {kernel,
     [{connect_all,true},
      {inet_default_connect_options,[{nodelay,true}]},
      {inet_dist_listen_max,25672},
      {inet_dist_listen_min,25672},
      {logger,
          [{handler,default,logger_std_h,
               #{config => #{type => standard_io},
                 formatter =>
                     {logger_formatter,
                         #{single_line => false,legacy_header => true}}}}]},
      {logger_level,notice},
      {logger_sasl_compatible,false},
      {net_tickintensity,4},
      {net_ticktime,60},
      {prevent_overlapping_partitions,false},
      {shell_docs_ansi,auto},
      {shell_history_drop,[]},
      {shutdown_func,{rabbit_prelaunch,shutdown_func}}]},
 {khepri,[{default_store_id,rabbitmq_metadata},{default_timeout,30000}]},
 {khepri_mnesia_migration,[]},
 {mnesia,
     [{dir,"/var/lib/rabbitmq/mnesia/rabbit@RabbitMQ"},
      {dump_log_time_threshold,90000},
      {dump_log_write_threshold,5000}]},
 {oauth2_client,[]},
 {observer_cli,
     [{plugins,
          [#{module => rabbit_observer_cli_classic_queues,title => "Classic",
             shortcut => "CQ",sort_column => 4},
           #{module => rabbit_observer_cli_quorum_queues,title => "Quorum",
             shortcut => "QQ",sort_column => 4}]},
      {scheduler_usage,disable}]},
 {os_mon,
     [{start_cpu_sup,false},
      {start_disksup,false},
      {start_memsup,false},
      {start_os_sup,false}]},
 {osiris,[{data_dir,"/var/lib/rabbitmq/mnesia/rabbit@RabbitMQ/stream"}]},
 {public_key,[]},
 {ra,[{data_dir,"/var/lib/rabbitmq/mnesia/rabbit@RabbitMQ/quorum"},
      {wal_max_batch_size,4096},
      {wal_max_size_bytes,536870912}]},
 {rabbit,
     [{auth_backends,[rabbit_auth_backend_internal]},
      {auth_mechanisms,['PLAIN','AMQPLAIN']},
      {autocluster,
          [{peer_discovery_backend,rabbit_peer_discovery_classic_config}]},
      {autoheal_state_transition_timeout,60000},
      {background_gc_enabled,false},
      {background_gc_target_interval,60000},
      {backing_queue_module,rabbit_priority_queue},
      {channel_max,2047},
      {channel_operation_timeout,15000},
      {channel_tick_interval,60000},
      {cluster_keepalive_interval,10000},
      {cluster_nodes,{[],disc}},
      {cluster_partition_handling,ignore},
      {collect_statistics,fine},
      {collect_statistics_interval,5000},
      {config_entry_decoder,[{passphrase,undefined}]},
      {consumer_timeout,1800000},
      {credentials_obfuscation_fallback_secret,<<"nocookie">>},
      {credit_flow_default_credit,{400,200}},
      {data_dir,"/var/lib/rabbitmq/mnesia/rabbit@RabbitMQ"},
      {dead_letter_worker_consumer_prefetch,32},
      {dead_letter_worker_publisher_confirm_timeout,180000},
      {default_consumer_prefetch,{false,0}},
      {default_permissions,[<<".*">>,<<".*">>,<<".*">>]},
      {default_user,<<"guest">>},
      {default_user_tags,[administrator]},
      {default_vhost,<<"/">>},
      {delegate_count,16},
      {disk_free_limit,50000000},
      {disk_monitor_failure_retries,10},
      {disk_monitor_failure_retry_interval,120000},
      {enabled_plugins_file,"/etc/rabbitmq/enabled_plugins"},
      {feature_flags_file,
          "/var/lib/rabbitmq/mnesia/rabbit@RabbitMQ-feature_flags"},
      {fhc_read_buffering,false},
      {fhc_write_buffering,true},
      {frame_max,131072},
      {halt_on_upgrade_failure,true},
      {handshake_timeout,10000},
      {heartbeat,60},
      {loopback_users,[<<"guest">>]},
      {max_message_size,134217728},
      {memory_monitor_interval,2500},
      {mirroring_flow_control,true},
      {mirroring_sync_batch_size,4096},
      {mnesia_table_loading_retry_limit,10},
      {mnesia_table_loading_retry_timeout,30000},
      {msg_store_credit_disc_bound,{4000,800}},
      {msg_store_file_size_limit,16777216},
      {msg_store_index_module,rabbit_msg_store_ets_index},
      {msg_store_io_batch_size,4096},
      {msg_store_shutdown_timeout,600000},
      {num_ssl_acceptors,10},
      {num_tcp_acceptors,10},
      {password_hashing_module,rabbit_password_hashing_sha256},
      {plugins_dir,
          "/usr/lib/rabbitmq/plugins:/usr/lib/rabbitmq/lib/rabbitmq_server-3.13.4/plugins"},
      {plugins_expand_dir,
          "/var/lib/rabbitmq/mnesia/rabbit@RabbitMQ-plugins-expand"},
      {proxy_protocol,false},
      {queue_explicit_gc_run_operation_threshold,1000},
      {queue_index_embed_msgs_below,4096},
      {queue_index_max_journal_entries,32768},
      {quorum_cluster_size,3},
      {quorum_commands_soft_limit,32},
      {ranch_connection_max,infinity},
      {release_series_eol_date,none},
      {reverse_dns_lookups,false},
      {server_properties,[]},
      {ssl_allow_poodle_attack,false},
      {ssl_apps,[asn1,crypto,public_key,ssl]},
      {ssl_cert_login_from,distinguished_name},
      {ssl_handshake_timeout,5000},
      {ssl_listeners,[]},
      {ssl_options,[]},
      {stream_messages_soft_limit,256},
      {tcp_listen_options,
          [{backlog,128},
           {nodelay,true},
           {linger,{true,0}},
           {exit_on_close,false}]},
      {tcp_listeners,[{"auto",5672}]},
      {trace_vhosts,[]},
      {track_auth_attempt_source,false},
      {tracking_execution_timeout,15000},
      {vhost_process_reconciliation_enabled,true},
      {vhost_process_reconciliation_run_interval,30},
      {vhost_restart_strategy,continue},
      {vm_memory_calculation_strategy,rss},
      {vm_memory_high_watermark,0.4},
      {vm_memory_high_watermark_paging_ratio,0.5},
      {writer_gc_threshold,1000000000}]},
 {rabbit_common,[]},
 {rabbitmq_management,
     [{content_security_policy,
          "script-src 'self' 'unsafe-eval' 'unsafe-inline'; object-src 'self'"},
      {cors_allow_origins,[]},
      {cors_max_age,1800},
      {http_log_dir,none},
      {load_definitions,none},
      {management_db_cache_multiplier,5},
      {max_http_body_size,10000000},
      {process_stats_gc_timeout,300000},
      {stats_event_max_backlog,250}]},
 {rabbitmq_management_agent,
     [{rates_mode,basic},
      {sample_retention_policies,
          [{global,[{605,5},{3660,60},{29400,600},{86400,1800}]},
           {basic,[{605,5},{3600,60}]},
           {detailed,[{605,5}]}]}]},
 {rabbitmq_mqtt,
     [{allow_anonymous,true},
      {default_user,<<"guest">>},
      {exchange,<<"amq.topic">>},
      {mailbox_soft_limit,200},
      {max_packet_size_authenticated,268435455},
      {max_packet_size_unauthenticated,65536},
      {max_session_expiry_interval_seconds,86400},
      {num_ssl_acceptors,10},
      {num_tcp_acceptors,10},
      {prefetch,10},
      {proxy_protocol,false},
      {retained_message_store,rabbit_mqtt_retained_msg_store_dets},
      {retained_message_store_dets_sync_interval,2000},
      {sparkplug,false},
      {ssl_cert_login,false},
      {ssl_listeners,[]},
      {tcp_listen_options,
          [{backlog,128},
           {nodelay,true},
           {send_timeout,15000},
           {send_timeout_close,true}]},
      {tcp_listeners,[1883]},
      {topic_alias_maximum,16},
      {vhost,<<"/">>}]},
 {rabbitmq_prelaunch,[]},
 {rabbitmq_web_dispatch,[]},
 {rabbitmq_web_mqtt,
     [{cowboy_opts,[]},
      {num_ssl_acceptors,10},
      {num_tcp_acceptors,10},
      {proxy_protocol,false},
      {ssl_config,[]},
      {tcp_config,[{port,15675}]}]},
 {ranch,[]},
 {recon,[]},
 {redbug,[]},
 {runtime_tools,[{ttb_autostart_module,ttb_autostart}]},
 {sasl,[{errlog_type,error}]},
 {seshat,[]},
 {ssl,[]},
 {stdlib,[]},
 {stdout_formatter,[]},
 {syntax_tools,[]},
 {sysmon_handler,
     [{busy_dist_port,true},
      {busy_port,false},
      {gc_ms_limit,0},
      {heap_word_limit,0},
      {port_limit,100},
      {process_limit,100},
      {schedule_ms_limit,0}]},
 {systemd,
     [{auto_formatter,true},
      {logger,
          [{handler,systemd_journal,systemd_journal_h,
               #{formatter => {logger_formatter,#{template => [msg]}}}}]},
      {unset_env,true},
      {watchdog_scale,2}]},
 {thoas,[]},
 {tools,[{file_util_search_methods,[{[],[]},{"ebin","esrc"},{"ebin","src"}]}]},
 {xmerl,[]}]

**Reply by tsigcn** [reply]
RabbitMQ version is 3.13.4 and log as following:

2024-07-26 13:59:47.856348+00:00 [info] <0.129812.0> Accepted MQTT connection 20.2.216.247:52336 -> 10.2.0.4:1883 for client ID MQTT_FX_Client
2024-07-26 14:00:22.635367+00:00 [info] <0.129812.0> MQTT connection <<"20.2.216.247:52336 -> 10.2.0.4:1883">> will terminate because peer closed TCP connection
2024-07-26 14:01:55.247447+00:00 [info] <0.129824.0> Accepted MQTT connection 20.2.216.247:52367 -> 10.2.0.4:1883 for client ID t9FJMWdGS-nJ22zXRCdYXg
2024-07-26 14:01:55.249872+00:00 [info] <0.129824.0> MQTT connection <<"20.2.216.247:52367 -> 10.2.0.4:1883">> will terminate because peer closed TCP connection

**Reply by tsigcn** [reply]
@michaelklishin I think this is caused by the client has received PUBACK instead of PUBREC.

As mentioned before:

[MQTT Publishes at QOS2 causes a connection failure for the publisher.](https://github.com/rabbitmq/rabbitmq-mqtt/issues/155)

**Reply by michaelklishin** [reply]
That issue is from 2017. The MQTT plugin in 3.13 is [very different](https://www.rabbitmq.com/blog/2023/03/21/native-mqtt).

**Reply by michaelklishin** [reply]
@ansd would know if this is an expected behavior but I can imagine that some tools would not handle a QoS downgrade gracefully. Many clients do, though, we have a good decade of evidence.

**Reply by michaelklishin** [reply]
I suspect that the answer to the question here is "no, there are no workarounds other than using QoS 1 which is a perfectly safe option in practice for an absolute majority of scenarios". Again, as more than a decade of evidence with several other protocols that use [publisher confirms](https://www.rabbitmq.com/docs/confirms) or its variation under a different name.

**Comment by michaelklishin** [comment]
I can imagine that some tools would not handle a QoS downgrade gracefully. Many clients do, though, we have a good decade of evidence.

I suspect that the answer to the question here is "no, there are no workarounds other than using QoS 1 which is a perfectly safe option in practice for an absolute majority of scenarios". Again, as more than a decade of evidence with several other protocols that use [publisher confirms](https://www.rabbitmq.com/docs/confirms) or its variation under a different name.

**Comment by ansd** [comment, marked as answer]
QoS 2 is unsupported in RabbitMQ, see https://www.rabbitmq.com/docs/mqtt#limitations

When a client sends a PUBLISH packet with QoS level 2, RabbitMQ will
* downgrade to QoS 1 in MQTT 3.1 and 3.1.1
* close the connection in MQTT 5.0.

Note that RabbitMQ implements the MQTT 5.0 behaviour according to the MQTT 5.0 spec:
> If the Server included a Maximum QoS in its CONNACK response to a Client and it receives a PUBLISH packet with a QoS greater than this, then it uses DISCONNECT with Reason Code 0x9B (QoS not supported) as described in [section 4.13](https://docs.oasis-open.org/mqtt/mqtt/v5.0/os/mqtt-v5.0-os.html#S4_13_Errors) Handling errors.

**Reply by tsigcn** [reply]
@ansd Would you please provide some information about what will happen with earlier verison of RabbitMQ which do not support MQTT 5.0 (.e.g 3.12.x)?

@michaelklishin I think the answer maybe "there are no workarounds other than using QoS 1 or  expect a next verison of RabbitMQ which support MQTT QoS 2" ? Is there a roadmap for MQTT QoS 2 support?

Thanks for your greatly help.

**Reply by michaelklishin** [reply]
@tsigcn David has already explained what will happen in versions earlier than 3.13: they only support MQTT3.1.1.

FYI, those versions are also all out of community support.

**Reply by michaelklishin** [reply]
We will not make any promises as to when or if QoS 2 may land in RabbitMQ. Like I said, QoS 1 is perfectly sufficient as far as data safety goes. QoS 2 does not deliver any meaningful safety improvements but takes months of effort to implement for a small team like ours.

Other protocols RabbitMQ supports do not have a direct QoS 2 equivalent and their users are perfectly fine because they do provide a confirmation mechanism not meaningfully different from QoS 1 in MQTT.

**Reply by tsigcn** [reply]
@michaelklishin Got it for information of earlier versions. 
Is there a RabbitMQ roadmap for MQTT QoS 2 support or not planned?

I think there are many scenes that clients send MQTT QoS 2 messages and cannot be changed to QoS 1 easily.
For example, our scenes have million of clients that cannot be changed to send QoS 1 message easily or need a long period.
Downgrade to QoS 1 is a good idea but it seems cannot be implement because the client got error and will keep retry.

Thanks for your patience.

**Reply by lukebakken** [reply]
> We will not make any promises as to when or if QoS 2 may land in RabbitMQ

@tsigcn - how is that statement not clear with regard to QoS 2? It is _not planned_, nor on any roadmap.

**Reply by tsigcn** [reply]
@michaelklishin @lukebakken 
It is clear with statement regard to QoS 2.
Your reply is faster than my post. :)

I think there is no better solution for my scenes at this time, and I will try to make clients send QoS 1 messages.
Thanks and regards.

**Reply by michaelklishin** [reply]
It may be added one day for cases such as this one but I though I have made it very clear that this is not a trivial time investment for a small team like ours, and we **absolutely cannot make any public promises**.

**Comment by lukebakken** [comment]
@tsigcn as explained [here](https://github.com/rabbitmq/rabbitmq-server/discussions/11842#discussioncomment-10161309), RabbitMQ follows the MQTT 5.0 spec when QoS 2 is requested. If you would like to see RabbitMQ's behavior changed (to support a downgrade to QoS 1, for instance), you could implement it yourself, or you could contact Broadcom to become a [paying RabbitMQ user](https://www.rabbitmq.com/contact#paid-support), and make an official feature request.
